### ISSUE TITLE
Uyuni starvation, all SSH-push registered client shown as inactives

### ISSUE THREAD

**Issue by phibid** [uyuni-project/uyuni#3182]
Our Uyuni infra is quite little with 100 registered clients and increasing, however since some months we notice a starvation of the Uyuni server some time to time during which it is not able to check the status of its Salt SSH-push registered clients. All the registered servers are consequently shown as inactive (“System not checking in with Uyuni”) in the Web GUI.

Important to mention that “push via SSH” is the registration method used for all the registered servers and that the Uyuni proxies which are salt-minion clients are not impacted by this issue.

When the starvation begins, it starts with this error 500:

```
2021-01-22 02:32:00,873 [Thread-120370] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: Response code: 500
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.createSaltException(HttpAsyncClientImpl.java:145)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.access$000(HttpAsyncClientImpl.java:27)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:121)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:101)
        at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:123)
        at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:181)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:442)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:332)
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
        ... 1 more
```

No more info in any logs, and repository syncs is still working without issue, so not seems to be an issue with Taskomatic. SSHPush tasks queue blocked at some point ? We did not had this issue with less than 50 registered servers if I remember correctly, so perhaps by tuning some parameters?

However I have tried to follow this https://www.uyuni-project.org/uyuni-docs/uyuni/large-deployments/tuning.html, but from what is mentioning the doc, it should not be necessary in my case (“Tuning is not required on installations of fewer than 1000 clients. Do not perform these instructions on small or medium scale installations.”), and no luck with it anyway.

A restart of Uyuni temporarily solves the issue.

Any ideas ? Somebody who experienced this same issue ?

Next time it occurs again (probably in the next few days), I can collect extra information if needed to try to spot what is going on.

And to finish, detailed version of the Uyuni server
```
# zypper info Uyuni-Server-release
Loading repository data...
Warning: Repository 'Main Update Repository' appears to be outdated. Consider using a different mirror or server.
Reading installed packages...


Information for package Uyuni-Server-release:
---------------------------------------------
Repository     : Uyuni Server Stable
Name           : Uyuni-Server-release
Version        : 2020.11-99.2.uyuni
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes
Status         : up-to-date
Source package : Uyuni-Server-release-2020.11-99.2.uyuni.src
Summary        : Uyuni Server
Description    : 
    Uyuni lets you efficiently manage physical, virtual,
    and cloud-based Linux systems. It provides automated and cost-effective
    configuration and software management, asset management, and system
    provisioning.
```

**Comment by phibid** [comment]
Not sure it makes a difference, but I forgot to mention that all (or at least 99%) of the registered server are going through Uyuni proxies:

```
# zypper info Uyuni-Proxy-release
Refreshing service 'spacewalk'.
Loading repository data...
Reading installed packages...


Information for package Uyuni-Proxy-release:
--------------------------------------------
Repository     : Uyuni Proxy Stable for openSUSE Leap 15.2 (x86_64)
Name           : Uyuni-Proxy-release
Version        : 2020.11-99.2.uyuni
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes
Status         : up-to-date
Source package : Uyuni-Proxy-release-2020.11-99.2.uyuni.src
Summary        : Uyuni Proxy
Description    : 
    Uyuni Proxies extend large and/or geographically
    dispersed Uyuni environments to reduce load on the Uyuni
    Server, lower bandwidth needs, and provide faster local
    updates.
```

**Comment by phibid** [comment]
I think I now finally understand what is happening. 

The scenario is the following:

1- Package refresh attempt done on a registered server through taskomatic/ssh-push
2- Seems that the Python script executed on the registered server get stuck and the ssh connection stays indefinitely opened:

```
salt     19870 19868  0 15:27 pts/10   00:00:00 ssh my_registered_server -o KbdInteractiveAuthentication=no -o PasswordAuthentication=no -o GSSAPIAuthentication=no -o ConnectTimeout=180 -o Port=22 -o IdentityFile=/srv/susemanager/salt/salt_ssh/mgr_ssh_id -o User=root /bin/sh  << 'EOF' set -e set -u DEBUG="" if [ -n "$DEBUG" ] then set -x fi SUDO="" if [ -n "" ] then SUDO="sudo " fi SUDO_USER="" if [ "$SUDO" ] && [ "$SUDO_USER" ] then SUDO="sudo -u " elif [ "$SUDO" ] && [ -n "$SUDO_USER" ] then SUDO="sudo " fi EX_PYTHON_INVALID=10 PYTHON_CMDS="python3 /usr/libexec/platform-python python27 python2.7 python26 python2.6 python2 python" for py_cmd in $PYTHON_CMDS do if command -v "$py_cmd" >/dev/null 2>&1 && "$py_cmd" -c "import sys; sys.exit(not (sys.version_info >= (2, 6)));" then py_cmd_path=`"$py_cmd" -c 'from __future__ import print_function;import sys; print(sys.executable);'` cmdpath=`command -v $py_cmd 2>/dev/null || which $py_cmd 2>/dev/null` if file $cmdpath | grep "shell script" > /dev/null then ex_vars="'PATH', 'LD_LIBRARY_PATH', 'MANPATH', \ 'XDG_DATA_DIRS', 'PKG_CONFIG_PATH'" export `$py_cmd -c \ "from __future__ import print_function; import sys; import os; map(sys.stdout.write, ['{0}={1} ' \ .format(x, os.environ[x]) for x in [$ex_vars]])"` exec $SUDO PATH=$PATH LD_LIBRARY_PATH=$LD_LIBRARY_PATH \ MANPATH=$MANPATH XDG_DATA_DIRS=$XDG_DATA_DIRS \ PKG_CONFIG_PATH=$PKG_CONFIG_PATH \ "$py_cmd_path" -c \ 'import base64; exec(base64.b64decode("""IyB ........
```

3 - Knowing that `taskomatic.ssh_push_workers` is set by default to 2 (https://www.uyuni-project.org/uyuni-docs/uyuni/client-configuration/contact-methods-pushssh.html), no more free ssh_push worker, and taskomatic can't launch more ssh-push connections. The following warning then continuously appears in the log:

```
2021-01-29 15:42:00,069 [DefaultQuartzScheduler_Worker-4] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-01-29 15:43:00,054 [DefaultQuartzScheduler_Worker-2] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-01-29 15:44:00,057 [DefaultQuartzScheduler_Worker-8] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-01-29 15:45:00,049 [DefaultQuartzScheduler_Worker-14] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
```

I have tried to increase `taskomatic.ssh_push_workers` from 2 to 8, but it does not help, as a retry is done every 10 minutes it seems, ending up by the lock of all the ssh_push workers.

3 thoughts:

1. It could perhaps be a good idea to encapsulate the remote Python script execution into a timeout shell command ?

2. Force a recycle of the ssh_push workers every now and then ?

3. I think that the default value of 2 for `taskomatic.ssh_push_workers` is far to be optimum.
    Defaulting it at least to 4 will help to speed up the refresh process, especially on a env with more than 100 registered servers.

**Comment by phibid** [comment]
Seems that lots of ssh_push workers are stuck because of this error which appears in the taskomatic log:

```
2021-02-01 17:27:02,518 [Thread-2668] WARN  com.suse.manager.webui.services.impl.SaltService - JsonParsingError({"stdout":"","stderr":"Traceback (most recent call last):\n  File \"/var/tmp/.root_930115_salt/salt-call\", line 26, in <module>\n    from salt.scripts import salt_call\n  File \"/var/tmp/.root_930115_salt/pyall/salt/scripts.py\", line 21, in <module>\n    from salt.exceptions import SaltSystemExit, SaltClientError, SaltReqTimeoutError\nModuleNotFoundError: No module named 'salt.exceptions'","retcode":1,"id":"my_registered_server","fun":"test.ping","jid":"20210201172700783697","_stamp":"2021-02-01T17:27:02.507327"}, No salt ssh return value, return code: 1)
```

And indeed, seems that the salt-ssh SHIM seems to be corrupted on all my ssh-push registered server since the last upgrade done on Uyuni server:

```
# salt-ssh --priv /srv/susemanager/salt/salt_ssh/mgr_ssh_id 'my_registered_server' test.ping
my_registered_server:
    ----------
    retcode:
        1
    stderr:
        Traceback (most recent call last):
          File "/var/tmp/.root_930115_salt/salt-call", line 26, in <module>
            from salt.scripts import salt_call
          File "/var/tmp/.root_930115_salt/pyall/salt/scripts.py", line 21, in <module>
            from salt.exceptions import SaltSystemExit, SaltClientError, SaltReqTimeoutError
        ModuleNotFoundError: No module named 'salt.exceptions'
    stdout:
```

Removing `/var/tmp/.root_930115_salt` on the registered clients seems to solve the issue. Not sure if this is a known bug or if a bad thing happened during the upgrade from 2020.09 -> 2020.11

Is there a way to force a rewrite of the SHIM on all the servers from the Uyuni server ?

**Comment by phibid** [comment]
Finally all those issues I have solved with the salt--ssh clients do not really help me. Now, all seems to be clean, no more errors regarding the salt clients connexions, however the salt-api just continue to crash some time to time.

A complete example:

```
2021-02-04 05:09:00,411 [DefaultQuartzScheduler_Worker-49] INFO  com.redhat.rhn.taskomatic.task.SSHPush - In the queue: 7
2021-02-04 05:09:00,490 [Thread-38784] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server1
2021-02-04 05:09:00,504 [Thread-38784] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server1
2021-02-04 05:09:00,509 [Thread-38785] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server2
2021-02-04 05:09:00,510 [Thread-38785] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server2
2021-02-04 05:09:00,551 [Thread-38783] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server3
2021-02-04 05:09:00,552 [Thread-38783] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server3
2021-02-04 05:09:00,551 [Thread-38786] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server4
2021-02-04 05:09:00,558 [Thread-38786] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server4
2021-02-04 05:09:00,551 [Thread-38771] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server5
2021-02-04 05:09:00,560 [Thread-38771] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server5
2021-02-04 05:09:00,616 [Thread-38781] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server6
2021-02-04 05:09:00,617 [Thread-38781] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server6
2021-02-04 05:09:00,659 [Thread-38782] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: server7
2021-02-04 05:09:00,659 [Thread-38782] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: server7
2021-02-04 05:09:01,142 [Thread-38785] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: Response code: 500
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.createSaltException(HttpAsyncClientImpl.java:145)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.access$000(HttpAsyncClientImpl.java:27)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:121)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:101)
        at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:123)
        at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:181)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:442)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:332)
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
        ... 1 more
2021-02-04 05:10:00,427 [DefaultQuartzScheduler_Worker-55] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-02-04 05:11:00,300 [DefaultQuartzScheduler_Worker-50] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-02-04 05:12:00,279 [DefaultQuartzScheduler_Worker-17] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
2021-02-04 05:13:00,309 [DefaultQuartzScheduler_Worker-63] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.
```

So knowing that `worker_threads` is set to 8 in `/etc/salt/master.d/susemanager.conf`, taskomatic can launch 8 ssh-push jobs in parallel. However and don't know why, salt-api sometimes crash with an error code 500, and from there the workers are stuck. The only way to come back in a clean situation is to restart at least taskomatic :-(

I have seen which seems to be a similar issue here:
  https://github.com/saltstack/salt/issues/23408
or here https://github.com/saltstack/salt/issues/23408

But not really found a solution.

**Comment by hustodemon** [comment]
Hi @phibid,

a couple of questions:

1. Which OS do run the affected clients?
1. Does the pinging of the affected clients work via the web UI? Just go to `Salt -> Remote Commands` and try to run `test.ping` on your salt-ssh clients.

**Comment by phibid** [comment]
Sorry for the mess, but I have mixed up 2 issues here, erroneously thinking that the issue 1 was the root cause of the issue 2:

**Issue 1:**

  **-Salt SHIM corrupted:** Following error appearing by the execution of a test.ping on any salt-ssh registered OS client (CentOS 7/8, Ubuntu 18.04/20.04):

```
# salt-ssh --priv /srv/susemanager/salt/salt_ssh/mgr_ssh_id 'my_registered_server' test.ping
my_registered_server:
    ----------
    retcode:
        1
    stderr:
        Traceback (most recent call last):
          File "/var/tmp/.root_930115_salt/salt-call", line 26, in <module>
            from salt.scripts import salt_call
          File "/var/tmp/.root_930115_salt/pyall/salt/scripts.py", line 21, in <module>
            from salt.exceptions import SaltSystemExit, SaltClientError, SaltReqTimeoutError
        ModuleNotFoundError: No module named 'salt.exceptions'
    stdout:
```

I have definitively solved this by manually removing the `/var/tmp/.root_930115_salt` directory to force a total re-creation of the SHIM on all the affected salt-ssh clients. I have noticed that the modification date of the directory `/var/tmp/.root_930115_salt` corresponds to the day I have upgraded Uyuni (2020.09 -> 2020.11). No idea about what's happened. No sure if you want me to open a separate bug for this, knowing that I would be unable to reproduce the error. However I would be able to give you the content of a "corrupted" `/var/tmp/.root_930115_salt`, as I have backup 1 in case of ....

I would suggest to improve the behaviour of Uyuni in those cases. If test.ping returns an exit code != 0, the affected registered client should be immediately flag in failed state.

**Issue 2:**

  **-Salt-api crash:** From time to time, error 500 appearing in the Taskomatic logs:

```2021-02-04 05:09:01,142 [Thread-38785] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: Response code: 500
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.createSaltException(HttpAsyncClientImpl.java:145)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.access$000(HttpAsyncClientImpl.java:27)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:121)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:101)
        at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:123)
        at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:181)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:442)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:332)
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
```

Direct consequence is the inability of Uyuni lo launch more test.ping actions:

```2021-02-04 05:10:00,427 [DefaultQuartzScheduler_Worker-55] WARN  com.redhat.rhn.taskomatic.task.SSHPush - Maximum number of workers already put ... skipping.```

Which leads to all the salt-ssh clients to be flagged as inactive after some few days. A restart of the Uyuni server solves the issue, but I guess that a restart of salt-api and/or salt-master should be enough, will try when it will crash again.

**Comment by meaksh** [comment]
This is really interesting. I've also noticed these exact same behaviors on a scenario where Salt SSH minions are also heavily used:

- **Issue 1:** Corrupted Salt "thin" deployed on the SSH minions, producing this exact exception:
```
        Traceback (most recent call last):
          File "/var/tmp/.root_930115_salt/salt-call", line 26, in <module>
            from salt.scripts import salt_call
          File "/var/tmp/.root_930115_salt/pyall/salt/scripts.py", line 21, in <module>
            from salt.exceptions import SaltSystemExit, SaltClientError, SaltReqTimeoutError
        ModuleNotFoundError: No module named 'salt.exceptions'
```

It is not yet clear to me what is causing the "thin" to get corrupted (with missing files). You mentioned this happened at the exact same time of an upgrade of Uyuni. We would need to investigate this a bit more.

**Question:** Is "salt-minion" also installed on those SSH minions somehow or are they pure SSH minions? 

Anyway, as a workaround to prevent this error from happening when the Salt "thin" gets corrupted in the SSH minion, you could create a cron job in the SSH minions that runs every minute or so to check the integrity of the "thin" and force the redeployed during next "salt-ssh" execution in such cases. Something like:

```bash
#!/bin/sh

if [ ! -f /var/tmp/.root_930115_salt/pyall/salt/exceptions.py ];
then
    echo "Detected corrupted Salt thin at /var/tmp/.root_930115_salt/ !!!"
    echo "Removing /var/tmp/.root_930115_salt/code-checksum to force thin redeployment"
    rm -f /var/tmp/.root_930115_salt/code-checksum || true;
fi
```

- **Issue 2:** Salt API returning 500 error from time to time:

I'm not sure if this is the same case, but we've also noticed that a race condition might happen at the time of Java generating the roster files for a Salt SSH action. When multiple SSH workers are running, it might be that the "salt-ssh" subprocess is started before the generated custom roster file has been synced on the filesystem by the Java stack. This could cause "salt-ssh" raise an error because the expected roster data is not there yet.

We have a tentative patch to avoid this issue, I'll create a PR soon and add the reference here.

Thanks for reporting and providing your finding here @phibid !

Hth!

**Comment by phibid** [comment]
They are pure SSH minions, no salt-minion installed. Thanks for the script, if I can see this occurring again, I will test it.

Can't wait to see your PR :-)

**Comment by mcalmer** [comment]
There were rumors that a systemd tmp cleanup job randomly remove files from "tmp" dirs.

**Comment by meaksh** [comment]
@phibid this is the PR for preventing the race condition at the time of generating the roster files by the SSH workers: https://github.com/uyuni-project/uyuni/pull/3225

Hth!

**Comment by phibid** [comment]
So I confirm that in my case a simple restart of salt-api (`systemctl restart salt-api`) solves the issue as the frozen salt-ssh connections are then freed up.

The stack trace appearing in the Taskomatic logs while restarting salt-api:

```
2021-02-08 14:24:25,450 [Thread-49583] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: org.apache.http.ConnectionClosedException: Connection closed
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: org.apache.http.ConnectionClosedException: Connection closed
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.suse.salt.netapi.exception.SaltException: org.apache.http.ConnectionClosedException: Connection closed
        at com.suse.manager.webui.services.impl.SaltService.adaptException(SaltService.java:950)
        at com.suse.manager.webui.services.impl.SaltSSHService.callSyncSSHInternal(SaltSSHService.java:701)
        at com.suse.manager.webui.services.impl.SaltSSHService.callSyncSSH(SaltSSHService.java:198)
        at com.suse.manager.webui.services.impl.SaltSSHService.callSyncSSH(SaltSSHService.java:215)
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:803)
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:214)
        ... 7 more
Caused by: org.apache.http.ConnectionClosedException: Connection closed
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.endOfInput(HttpAsyncRequestExecutor.java:350)
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:261)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
        ... 1 more
```

**Comment by phibid** [comment]
The PR has been committed in 2021.11 right ?

I ask because no more crash until now, and those continuous messages I had before in the taskomatik logs are not appearing anymore

```
2021-02-06 05:45:00,045 [DefaultQuartzScheduler_Worker-31] INFO  com.redhat.rhn.taskomatic.task.SessionCleanup - 16 stale session(s) deleted
```

**Comment by meaksh** [comment]
@phibid actually not, the PR was not included in `2021.01` release.

**Comment by phibid** [comment]
Ah oops, indeed, it just crashed:

```
2021-02-11 09:33:00,618 [Thread-30333] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: Response code: 500
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)

```

**Comment by meaksh** [comment]
Hey @phibid,
did you get any feedback already after removing the potencial conflicting tuning at `rhn.conf` file?

Are those 500 errors still happening?

@vzhestkov has been also working fixing some Salt SSH issues that might be affecting also here. Let see how your feedback is and we will see.

Thanks!

**Comment by phibid** [comment]
Hi @meaksh !

Unfortunatly, after 12 days without a crash, it occurred again last night :-(

```
2021-05-11 01:05:01,051 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Generating new repository metadata for channel 'centos7-x86_64-cr'(sha1) 4715 packages, 434 errata
2021-05-11 01:05:14,134 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Processed 1000 packages
2021-05-11 01:05:25,177 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Processed 2000 packages
2021-05-11 01:05:33,930 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Processed 3000 packages
2021-05-11 01:05:45,011 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Processed 4000 packages
2021-05-11 01:06:00,113 [DefaultQuartzScheduler_Worker-5] WARN  com.redhat.rhn.taskomatic.task.ChannelRepodata - Maximum number of workers already put ... skipping.
2021-05-11 01:06:02,265 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Processed 4715 packages
2021-05-11 01:06:03,704 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Repository metadata generation for 'centos7-x86_64-cr' finished in 62 seconds
2021-05-11 01:06:04,324 [Thread-133243] INFO  com.redhat.rhn.taskomatic.task.repomd.RepositoryWriter - Solv file successfully create for 'centos7-x86_64-cr'
2021-05-11 01:15:00,116 [DefaultQuartzScheduler_Worker-17] INFO  com.redhat.rhn.taskomatic.task.SessionCleanup - 32 stale session(s) deleted
2021-05-11 01:20:00,689 [DefaultQuartzScheduler_Worker-13] INFO  com.redhat.rhn.taskomatic.task.SSHPush - In the queue: 6
2021-05-11 01:20:00,705 [Thread-133358] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: aaa
2021-05-11 01:20:00,706 [Thread-133358] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: bbb
2021-05-11 01:20:00,715 [Thread-133357] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Executing actions for minion: ccc
2021-05-11 01:20:00,716 [Thread-133357] INFO  com.redhat.rhn.taskomatic.task.SSHPush - Performing a check-in for: ddd
2021-05-11 01:20:00,855 [Thread-133358] ERROR com.redhat.rhn.taskomatic.task.SSHPush - com.suse.salt.netapi.exception.SaltException: Response code: 500
java.lang.RuntimeException: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.manager.webui.services.impl.SaltService.callSync(SaltService.java:228)
        at com.suse.manager.webui.services.impl.SaltService.ping(SaltService.java:237)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.performCheckin(SSHPushWorkerSalt.java:327)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.lambda$run$0(SSHPushWorkerSalt.java:112)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at com.redhat.rhn.taskomatic.task.sshpush.SSHPushWorkerSalt.run(SSHPushWorkerSalt.java:100)
        at EDU.oswego.cs.dl.util.concurrent.PooledExecutor$Worker.run(PooledExecutor.java:732)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.suse.salt.netapi.exception.SaltException: Response code: 500
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.createSaltException(HttpAsyncClientImpl.java:145)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl.access$000(HttpAsyncClientImpl.java:27)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:121)
        at com.suse.salt.netapi.client.impl.HttpAsyncClientImpl$1.completed(HttpAsyncClientImpl.java:101)
        at org.apache.http.concurrent.BasicFuture.completed(BasicFuture.java:123)
        at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.responseCompleted(DefaultClientExchangeHandlerImpl.java:181)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.processResponse(HttpAsyncRequestExecutor.java:442)
        at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.inputReady(HttpAsyncRequestExecutor.java:332)
        at org.apache.http.impl.nio.DefaultNHttpClientConnection.consumeInput(DefaultNHttpClientConnection.java:265)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:81)
        at org.apache.http.impl.nio.client.InternalIODispatch.onInputReady(InternalIODispatch.java:39)
        at org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady(AbstractIODispatch.java:121)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.readable(BaseIOReactor.java:162)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:337)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:315)
        at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:276)
        at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:104)
        at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:588)
        ... 1 more
```

**Comment by vzhestkov** [comment]
@phibid , Hi.
The error on salt-api side is more important in this case, but for some cases the messages with default logging level are not very useful. It's better to set logging level to `trace` for salt master and restart `salt-api`.

**Comment by phibid** [comment]
@vzhestkov Done, thanks ! Will let you know next time it crashs again.

**Comment by vzhestkov** [comment]
@phibid, you can upload the log somewhere for me to check, but note that it contains quite sensitive data. (passwords, hostnames etc.)

**Comment by phibid** [comment]
Got it !! It crahed last night at exactly 00h03, and I have spot this in the `/var/log/salt/api` log:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/netapi/rest_cherrypy/app.py", line 860, in hypermedia_handler
    ret = cherrypy.serving.request._hypermedia_inner_handler(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 54, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/usr/lib/python3.6/site-packages/salt/netapi/rest_cherrypy/app.py", line 2140, in POST
    'return': list(self.exec_lowstate()),
  File "/usr/lib/python3.6/site-packages/salt/netapi/rest_cherrypy/app.py", line 1187, in exec_lowstate
    ret = self.api.run(chunk)
  File "/usr/lib/python3.6/site-packages/salt/netapi/__init__.py", line 152, in run
    return l_fun(*f_call.get('args', ()), **f_call.get('kwargs', {}))
  File "/usr/lib/python3.6/site-packages/salt/netapi/__init__.py", line 218, in ssh
    return ssh_client.cmd_sync(kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/ssh/client.py", line 206, in cmd_sync
    **kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/ssh/client.py", line 168, in cmd
    **kwargs)
  File "/usr/lib/python3.6/site-packages/salt/client/ssh/client.py", line 115, in _prep_ssh
    ret = salt.client.ssh.SSH(opts)
  File "/usr/lib/python3.6/site-packages/salt/client/ssh/__init__.py", line 233, in __init__
    self._update_targets()
  File "/usr/lib/python3.6/site-packages/salt/client/ssh/__init__.py", line 408, in _update_targets
    if salt.utils.network.is_reachable_host(hostname):
  File "/usr/lib/python3.6/site-packages/salt/utils/network.py", line 251, in is_reachable_host
    assert type(socket.getaddrinfo(entity_name, 0, 0, 0, 0)) == list
  File "/usr/lib64/python3.6/socket.py", line 745, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
TypeError: getaddrinfo() argument 1 must be string or None
2021-05-16 00:03:00,487 [cherrypy.access.140022232809824:284 ][INFO    ][28501] 127.0.0.1 - - [16/May/2021:00:03:00] "POST /run HTTP/1.1" 500 57 "" "Apache-HttpAsyncClient/4.1.4 (Java/11.0.10)"
```

**Comment by vzhestkov** [comment]
@phibid, Hi. I think it's time for to migrate to new stable version of Uyuni. The issue was fixed with this PR: https://github.com/openSUSE/salt/pull/336

**Comment by phibid** [comment]
@vzhestkov Hi ! I already use the last Uyuni stable version:

```
# zypper info Uyuni-Server-release
Loading repository data...
Reading installed packages...


Information for package Uyuni-Server-release:
---------------------------------------------
Repository     : Uyuni Server Stable
Name           : Uyuni-Server-release
Version        : 2021.04-122.3.uyuni
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes
Status         : up-to-date
Source package : Uyuni-Server-release-2021.04-122.3.uyuni.src
Summary        : Uyuni Server
Description    : 
    Uyuni lets you efficiently manage physical, virtual,
    and cloud-based Linux systems. It provides automated and cost-effective
    configuration and software management, asset management, and system
    provisioning.
```

**Comment by vzhestkov** [comment]
What is the `salt-master` and `python3-salt` packages versions are installed?

**Comment by phibid** [comment]
```
# zypper info salt-master 
Loading repository data...
Reading installed packages...


Information for package salt-master:
------------------------------------
Repository     : Main Update Repository
Name           : salt-master
Version        : 3000-lp152.3.33.1
Arch           : x86_64
Vendor         : openSUSE
Installed Size : 3.0 MiB
Installed      : Yes (automatically)
Status         : up-to-date
Source package : salt-3000-lp152.3.33.1.src
Summary        : The management component of Saltstack with zmq protocol supported
Description    : 
    The Salt master is the central server to which all minions connect.
    Enabled commands to remote systems to be called in parallel rather
    than serially.
```

```
# zypper info python3-salt
Loading repository data...
Reading installed packages...


Information for package python3-salt:
-------------------------------------
Repository     : Main Update Repository
Name           : python3-salt
Version        : 3000-lp152.3.33.1
Arch           : x86_64
Vendor         : openSUSE
Installed Size : 44.7 MiB
Installed      : Yes (automatically)
Status         : up-to-date
Source package : salt-3000-lp152.3.33.1.src
Summary        : python3 library for salt
Description    : 
    Python3 specific files for salt
```

**Comment by vzhestkov** [comment]
It seems the fix was not included in this release.
Could you please check with updating salt packages from https://download.opensuse.org/repositories/home:/vzhestkov:/branches:/systemsmanagement:/saltstack:/products:/3000/openSUSE_Leap_15.2/
on Uyuni Server?

**Comment by phibid** [comment]
I do not have an Uyuni test server. I think that I will wait for the next Uyuni release to install those new Salt packages if you don't mind.

**Comment by vzhestkov** [comment]
I don't. In this case it's probably better to close the issue. Feel free to reopen it if you get this issue after migrating to the next Uyuni release.

**Comment by meaksh** [comment]
@phibid I'm closing this issue now since the fix for this "salt-ssh" crash is already done and you should get it as soon as the upgrade for Salt reaches Leap 15.2 repositories.

If you're facing this issue again after upgrading, please reopen this and we can continue debugging!

Thanks a lot for your collaboration here! :+1:
